src/par/FixColumnPartitions.c

```diff
@@ -19,134 +19,166 @@
 
-    rcen[0] = rcen[1] = rcen[2] = 0.0;
-    MR_Coords(mr,&nrv,rxyz);
-    for (i = 0; i < nrv; i++) {
-      for (k = 0; k < 3; k++)
-        rcen[k] += rxyz[i][k];
+   Returns 1 if the relationship is proper, or 0 if the region has
+   zero volume and is therefore uncertain.
+*/
+int FixColumnPartitions_UpDown(Mesh_ptr mesh, MRegion_ptr mr, MFace_ptr* up, MFace_ptr* dn) {
+  int found;
+  List_ptr rfaces, rfaces2;
+  int nrf, i, j, k, nfv, ret;
+  MFace_ptr rf; double rcen[3], upcen[3], dncen[3];
+  double fxyz[MAXPV2][3];
+  MRegion_ptr mr_it;
+
+  *up=NULL; *dn=NULL;
+  
+  rfaces = MR_Faces(mr);
+  nrf = List_Num_Entries(rfaces);
+
+  /* find two faces that are not sides */
+  found = 0; i = 0;
+  while (found < 2) {
+    rf = List_Entry(rfaces,i);
+    if (!FixColumnPartitions_IsSide(mesh, mr, rf)) {
+      if (found < 1) *up = rf;
+      else *dn = rf;
+      found++;
     }
+    i++;
+    if (i >= nrf) 
+      MSTK_Report("FixColumnPartitions","Mesh is not columnar, can't find both up and down faces.",MSTK_FATAL);
+  }
+
+  /* figure which is up and which is down */
+  MF_Coords(*up,&nfv,fxyz);
+  upcen[0] = upcen[1] = upcen[2] = 0.;
+  for (j = 0; j < nfv; j++) {
     for (k = 0; k < 3; k++)
-      rcen[k] /= nrv;        
-
-    topf = NULL;
-    maxdp = -1.0e+10;
-    for (i = 0; i < nrf; i++) {
-      rf = List_Entry(rfaces,i);
-      MF_Coords(rf,&nfv,fxyz);
-
-      fcen[0] = fcen[1] = fcen[2] = 0.0;
-      for (j = 0; j < nfv; j++) {
-        for (k = 0; k < 3; k++)
-          fcen[k] += fxyz[j][k];
-      }
-      for (k = 0; k < 3; k++)
-        fcen[k] /= nfv;
-
-      for (k = 0; k < 3; k++)
-        rfvec[k] = fcen[k]-rcen[k];
-      norm = sqrt(rfvec[0]*rfvec[0]+rfvec[1]*rfvec[1]+rfvec[2]*rfvec[2]);
-      for (k = 0; k < 3; k++)
-        rfvec[k] /= norm;
-
-      /* rfvec[2] is same as rfvec dotted with [0,0,1] */
-      if (rfvec[2] > maxdp) {
-        maxdp = rfvec[2];
-        topf = rf;
-      }
-    }
-    List_Delete(rfaces);
+      upcen[k] += fxyz[j][k];
+  }
 
-    if (!topf)
-      MSTK_Report("FixColumnPartitions","Could not find top face?",MSTK_FATAL);
+  MF_Coords(*dn,&nfv,fxyz);
+  dncen[0] = dncen[1] = dncen[2] = 0.;
+  for (j = 0; j < nfv; j++) {
+    for (k = 0; k < 3; k++)
+      dncen[k] += fxyz[j][k];
+  }
 
-    /* Check if the top face is also a top boundary face */
+  /* ASSERT x,y are the same */
+  if (fabs(upcen[0] - dncen[0]) > 1.e-6 ||
+      fabs(upcen[1] - dncen[1]) > 1.e-6) {
+    MSTK_Report("FixColumnPartitions","Mesh is not quite columnar, up and down faces do not align.",MSTK_FATAL);
+  }
 
-    fregs = MF_Regions(topf);
-    nfr = List_Num_Entries(fregs);
+  if (upcen[2] < dncen[2] - 1.e-6) {
+    rf = *up;
+    *up = *dn;
+    *dn = rf;
+    ret = 1;
+  } else if (upcen[2] > dncen[2] + 1.e-6) {
+    ret = 1;
+  } else {
+    /* find the first region upward that is proper */
+    MSTK_Report("FixColumnPartitions","Partitioning with degenerate cells is not yet supported.",MSTK_FATAL);
```

Comment:
Its the comparing of relative elevation of face centers that is leading you to this ambiguity. Instead if you found the face that most aligned with the +ve z direction and the face that most aligned with the -ve z direction (after getting the face normal such that its pointing out of the region), you would not have any ambiguity.